### ISSUE TITLE
ScoreTensorFlowModel does not produce the expected results

### ISSUE THREAD

**Issue by ileve** [dotnet/machinelearning#2462]
### System information

- **OS version/distro**: Win10-x64 10.0.17134
- **.NET Version (eg., dotnet --info)**: dotnet SDK 2.1.402

visual Studio 15.8.4

### Issue
`ScoreTensorFlowModel` does not produce the expected results from a pre-trained network
- **What did you do?**
I trained a [regression model](https://colab.research.google.com/drive/1kuR3ugh-EOXcNXz_Qglt06MQQQkA2Mmp) and saved the full model as an h5 file. Then i load the file and convert it to a pb file. Then the pb file was fed to an ML.NET project.

- **What happened?**
The predicted results from ML.NET project are not the same with the predicted results in TensorFlow notebook.
- **What did you expect?**
to get the same results.
### Source code / logs
the C# code was:

```csharp    
    class Program
    {
        static string dataset = Path.Combine(Directory.GetCurrentDirectory(), "mpg.txt");
        static string pbModel = Path.Combine(Directory.GetCurrentDirectory(), "fromh5.pb");

        static void Main(string[] args)
        {
            var mlContext = new MLContext(); 

            string dataFile = File.ReadAllText(dataset);
            // Creating a data reader, based on the format of the data
            var reader = TextLoaderStatic.CreateReader(mlContext, c => (
                        index: c.LoadFloat(0),  //it is irrelevant with inputs or output
                        dense_input_1: c.LoadFloat(1, 9)),
                        separator: '\t', hasHeader: true);

            // Read the data
            var data = reader.Read(new MultiFileSource(dataset));

            var estimator = mlContext.Transforms.ScoreTensorFlowModel(
                                    pbModel,
                                    new string[] { "dense_2_1/BiasAdd" },
                                    new string[] { "dense_input_1" });

            var model = estimator.Fit(data.AsDynamic);
            var predictionMachine = model.CreatePredictionEngine<InputData, OutputData>(mlContext);
            var inputs = mlContext.CreateEnumerable<InputData>(data.AsDynamic, reuseRowObject: false).ToArray();
            for (int i = 0; i < inputs.Length; i++)
            {
                var predictedLabel = predictionMachine.Predict(inputs[i]);
                for (int j = 0; j < inputs[i].Features.Length; j++)
                {
                    Console.Write(inputs[i].Features[j]);
                    Console.Write(" ");
                }
                Console.WriteLine(predictedLabel.Output[0]);
            }

        }
    }

    class InputData
    {
        [ColumnName("dense_input_1"), VectorType(9)]
        public float[] Features { get; set; }
    }

    class OutputData
    {
        [ColumnName("dense_2_1/BiasAdd"), VectorType(1)]
        public float[] Output { get; set; }
    }
```

The inputs which were used for prediction were read from `mpg.txt` file with contents:

```	
         Cylinders 	Displacement 	Horsepower 	Weight 	Acceleration 	Model Year 	USA 	Europe 	Japan
9 	1.483887	 1.865988 	2.234620	1.018782	-2.530891 	-1.604642 	0.774676 	-0.465148 	-0.495225
25 	1.483887 	1.578444 	2.890853 	1.925289 	-0.559020 	-1.604642 	0.774676 	-0.465148 	-0.495225
28 	1.483887 	1.041693 	2.313368 	2.063931 	1.054328 	-1.604642 	0.774676 	-0.465148 	-0.495225
31 	-0.869348 	-0.789008 	-0.259066 	-0.903250 	-0.559020 	-1.332580 	-1.286751 	-0.465148 	2.012852
33 	0.307270 	0.351586 	-0.127819 	-0.422150 	-0.917542 	-1.332580 	0.774676 	-0.465148 	-0.495225
39 	1.483887 	1.961837 	1.840880 	1.746357 	-1.455325 	-1.332580 	0.774676 	-0.465148 	-0.495225
40 	1.483887 	1.492180 	1.263395 	1.379015 	-0.738281 	-1.332580 	0.774676 	-0.465148 	-0.495225
43 	1.483887 	1.961837 	1.709634 	2.080521 	-1.276064 	-1.332580 	0.774676 	-0.465148 	-0.495225
48 	0.307270 	0.524113 	-0.442811 	0.176263 	-0.379759 	-1.332580 	0.774676 	-0.465148 	-0.495225
54 	-0.869348 	-1.181986 	-0.941548 	-1.632011 	0.875068 	-1.332580 	-1.286751 	-0.465148 	2.012852
```

The pb file is:
[pbfile](https://user-images.githubusercontent.com/31559543/52429060-05e4e680-2b0c-11e9-8a15-59e04c24ffc8.jpg)

Predicted results :
```15.770189 ,  8.624239 ,  7.2484765, 24.726585 , 21.833643 , 12.528247 , 13.377004 , 12.101825 , 20.065805 , 32.288208```

Predicted results from ML.NET:
```-0.5800864, -0.775505, -0.7530465, -0.1452617, -0.2986267, -0.6405546, -0.4707591, -0.6763195, -0.1432758, -0.2586956```

**Comment by Ivanidzo4ka** [comment]
Thank you for reporting this.  Would it be hard for you to share model itself? I know you posted link to it, but it required permission to access it.

**Comment by ileve** [comment]
Check this [link on github](https://github.com/ileve/debug-netcoreapp2.0/blob/master/Copy_of_basic_regression.ipynb)

**Comment by Ivanidzo4ka** [comment]
@zeahmed any chance you look on this one?

**Comment by zeahmed** [comment]
O.K.

**Comment by abgoswam** [comment]
@ileve could you try with latest ML.NET ?  also we would need the model if we want to discuss this further..

**Comment by ileve** [comment]
Specify the exact version of ML.NET and i will try to test it.
The C# model and the respective python tensorflow model are available in the previous posts above.
I you want something additionally, please be more specific.

**Comment by abgoswam** [comment]
@ileve . The latest version of ML.NET  v 1.0- Several  fixes have gone in since you  reported the issue in Feb . 

Also, a few clarifying questions:

1. Can u share the C# app  with the ML.NET v1.0 APIs that you are using to verify the scores match or not ? I want to make sure  the  pipelines match i.e. you are applying the appropriate transformations  before feeding data into the model 

2.  You are using keras  to first save model in h5 format, Then you are loading up the h5 model  and saving again in pb format.  Any particular reason for  using the h5 format as an intermediate step ?

**Comment by ileve** [comment]
I also need to update the following packages, which are required to execute the C# ML.Net project.
Microsoft.ML.StaticPipe
Microsoft.ML.TensorFlow

The latest versions are 0.12.0, i think. Do i reinstall these packages with version 0.12.0?

**Comment by abgoswam** [comment]
@ileve . yeah you would want to pick up the latest version of the Microsoft.ML.TensorFlow  nuget. 

Regarding Microsoft.ML.StaticPipe,  please note the Static API is considered experimental at this point.  
- The Ml.NET [cookbook](https://github.com/dotnet/machinelearning/blob/master/docs/code/MlNetCookBook.md) recommends using the "Dynamic API" paradigm.  
- If you are following the latest samples  for the v1.0 release of ML.NET then you would not need the Microsoft.ML.StaticPipe nuget at all

**Comment by ileve** [comment]
I have tested it again, using the new version 1.0.0 with dynamic API as you suggested, but i get the same faulty results as before.

I have uploaded the ML.NET project in the following link as you requested.
[ML.NET-TensorFlow model](https://github.com/ileve/ML.NET-TensorFlowModel)

**Comment by abgoswam** [comment]
@ileve  looking at the ipython [notebook ](https://github.com/ileve/debug-netcoreapp2.0/blob/master/Copy_of_basic_regression.ipynb)(that you used to build your model)  it seems you are using normalized values  to make the predictions.

However, in the corresponding ML.NET pipeline I do not see the same normalization being performed.  I believe that is leading to the discrepancy in the scores.

I am closing this issue. Please feel free to open a separate issue once you have ensured we are doing apples to apples comparison. 

@codemzs

**Comment by ileve** [comment]
I am sure that i was comparing apples to apples. The pipeline in ML.NET project does not contain any data transformation because i fed the model with the transformed values and not the raw data.

**Comment by abgoswam** [comment]
ok i will take a closer look at the dataset as well
